### ISSUE TITLE
[dask] Add support for early stopping in Dask interface

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#3712]
## Summary

`DaskLGBMClassifier` and `DaskLGBMRegressor` in the Python package should support early stopping.

## Motivation

Early stopping is generally useful with gradient boosting algorithms, to avoid wasted training iterations or unnecessary growth in the model size once desirable performance has been achieved. This feature is available in the non-Dask interfaces for LightGBM, and should be available with the Dask one.

## Description

This should mimic the approach XGBoost took (https://github.com/dmlc/xgboost/blob/516a93d25c3b6899558700430ffc99a29ea21e1a/python-package/xgboost/dask.py#L1386), where `eval_set` contains Dask collection s(Dask Array or Dask DataFrame).

## References

* relevant XGBoost PR (see linked issues and PRs): https://github.com/dmlc/xgboost/issues/5495
* one proposal: https://github.com/microsoft/LightGBM/pull/3515#issuecomment-752709132, https://github.com/h2oai/dask-lightgbm/commit/8cc8e83b0235050286a89dbbe2fd9d6b2e08271b

**Comment by jameslamb** [comment]
Closing in favor of being in #2302 with other feature requests. Please leave a comment here if you'd like to work on this.

**Comment by ffineis** [comment]
Hey I'd like to take this if it's cool - planning mainly base this off of [changes](https://github.com/h2oai/dask-lightgbm/commit/8cc8e83b0235050286a89dbbe2fd9d6b2e08271b) called out in #3515 and try to bring this in line with xgboost.dask.

**Comment by jameslamb** [comment]
sure, thank you! I'm really close to having a small reproducible example for the random "cannot bind to port XXXX" issue, will link that here when I've written it up.

**Comment by ffineis** [comment]
> sure, thank you! I'm really close to having a small reproducible example for the random "cannot bind to port XXXX" issue, will link that here when I've written it up.

Omigod lifesaver, thank you!

**Comment by jameslamb** [comment]
@ffineis do you think you'll have time to work on this this week? We're planning to do a 3.2.0 release of LightGBM in the next week or two. I didn't include this in my list of must-have Dask features for the next release  (https://github.com/microsoft/LightGBM/pull/3872#issuecomment-768794191), but I'd love to try to get this change in if we can since it can have such a big impact on training runtime.

If you don't have time this week, could I take this back from you and try it out?

Thanks so much for all your help with the Dask module so far!!

**Comment by ffineis** [comment]
Hey! Yes sorry I’ve started planning for this, gimme another week? The only
tricky part of matching xgboost.dask’s implementation is the id(training
data) check. So yeah, sorry haven’t made any progress yet but was planning
to make some commits this week.

**Comment by jameslamb** [comment]
oh yeah no problem, thanks!

**Comment by jmoralez** [comment]
Hi. I've been playing around with this and have a working (although horrible) implementation of this. The approach I took was using the futures of the persisted collections instead of turning the collections to lists of delayeds, so this avoids recomputing the training set in case it is in the `eval_set` as well (because both futures have the same key). Would you guys be interested in discussing this approach?

**Comment by jameslamb** [comment]
Interesting! I'll leave it to @ffineis to comment on that. Right now, I think our highest priority is supporting early stopping, and it would be ok if the first implementation of that merged to `master` computes the training data twice.

**Comment by ffineis** [comment]
Hey @jmoralez, thanks for the ideation! Honestly, don't feel bad, I think any implementation of ES will be pretty hairy given that we're using lists of delayed partitions instead of distributed lgbm.Datasets. I'm a fan of how xgboost.dask attempts to accomplish what you've mention via `id` - I'm just going to check whether each `X` in a (X, y) eval set shares the same `id` as that of `data`. If so, this will trigger within `_train_part` (local worker training function) just use the local `data` in for any eval set's parts that had originally called for `data`.

This method works if the entirety of an eval `X` equals `data`, but I'm guessing your approach goes even further in that it would also include cases when an eval `X` is a subset of partitions of training `data`? I'd say in the spirit of how PR's have been going (modular, one change at a time) - hold off until a naive early stopping approach gets merged, but then open an issue and follow up with an improvement PR - does that work?

**Comment by jmoralez** [comment]
Yeah that sounds fair. Is there a plan to create an `lgb.DaskDataset`?

**Comment by ffineis** [comment]
@jmoralez my guess is yes - I think check out #2302? Jlamb has a running list of dask features somewhere.

PR for this coming tomorrow night.

**Comment by jameslamb** [comment]
yes but I haven't written it up. Will do that right now. I like what `xgboost` did with `DaskDMatrix`. https://github.com/dmlc/xgboost/blob/9a0399e8981b2279d921fe2312f7ab1b880fd3c3/python-package/xgboost/dask.py#L185

**Comment by jameslamb** [comment]
I've added #3944 to track the idea of a `DaskDataset`. That's a fairly invasive change so I don't think it should be done before the 3.2.0 release (#3872)

**Comment by StrikerRUS** [comment]
Closing for now due to the lack of active work on this feature and open PRs.
